### DISCUSSION TITLE
Dynamic Fields on Entities

### DISCUSSION THREAD

**Discussion by rafaelmartinelli** [symfony/symfony#54608] [Q&A]
Hello all,

This is a DB + Symfony + EasyAdminBundle question. I have an entity called `Event`. An event has some boring static fields like `name`, `url`, `e-mail`, etc. Another entity, called `Participant`, is part of an event, and also has some boring static fields, like `name`, `e-mail`, etc. Here is the catch, when creating an event, the user may say he wants some custom fields filled by each participant for his event, for example, telephone number, institution, etc. So, I was thinking how to code something flexible in this case. My current idea is to create two helper classes `DynamicFieldTrait` and `DynamicFieldValue`, like the following (assume everything is in the DB using Doctrine ORM):

```php
class DynamicFieldType
{
    // To be used by DynamicFieldType::$type.
    public const FIELD_STRING = 0;
    public const FIELD_TEXT = 1;
    public const FIELD_RICH = 2;
    public const FIELD_INTEGER = 3;
    public const FIELD_NUMBER = 4;
    // (...)

    private int $id;
    private string $name;
    private int $type;
    private Event $event;
}

class DynamicFieldValue
{
    private int $id;
    private DynamicFieldType $type;
    private Participant $participant;
    private mixed $value;
}

class Event
{
    private int $id;
    private string $name;
    private string $url;
    private string $email;
    /** @var Participant[] $participants */
    private $participants;
    /** @var DynamicFieldType[] $fields */
    private $fields;
}

class Participant
{
    private int $id;
    private string $name;
    private string $email;
    private Event $event;
    /** @var DynamicFieldValue[] $values */
    private $values;
}
```

First question is, would it be the best design for what I want to achieve?

And then, I'm still thinking how to build a form for `Participant`, based on the fields described in `Event`, especially using EasyAdminBundle.

Thank you!

**Comment by chapterjason** [comment]
Are you familiar with the [Entity–attribute–value model](https://en.wikipedia.org/wiki/Entity%E2%80%93attribute%E2%80%93value_model)?

**Reply by rafaelmartinelli** [reply]
Hello, @chapterjason. Thanks for your reply! I am not familiar, but I've just read the Wikipedia page, and it makes sense. Would it be an EAV with two entities like below?

```php
// Entity 1
class Event
{
    private int $id;
    private string $name;
    private string $url;
    private string $email;
    /** @var Participant[] $participants */
    private $participants;
    /** @var DynamicFieldAttribute[] $attributes */
    private $attributes;
    /** @var DynamicFieldValue[] $values */
    private $values;
}

// Entity 2
class Participant
{
    private int $id;
    private string $name;
    private string $email;
    private Event $event;
    /** @var DynamicFieldValue[] $values */
    private $values;
}

// Attribute
class DynamicFieldAttribute
{
    // To be used by DynamicFieldAttribute::$type.
    public const ATTR_STRING = 0;
    public const ATTR_TEXT = 1;
    public const ATTR_RICH = 2;
    public const ATTR_INTEGER = 3;
    public const ATTR_NUMBER = 4;
    // (...)

    private int $id;
    private string $name;
    private int $type;
    private Event $event;
}

// Value
class DynamicFieldValue
{
    private int $id;
    private Event $event;
    private Participant $participant;
    private DynamicFieldAttribute $attribute;
    private mixed $value;
}
```

Would it be the best design using the EAV model? Maybe I don't need to link the `DynamicFieldValue` with `Event`, since the link exists through `DynamicFieldAttribute`...

Now I need to figure out how to generate a form for these values. I was thinking about a `CollectionField`, but each entry can be from a different type. Anyway, if the current design is ok, I can try for a while, and then come back with the results.

Thanks again!

**Comment by rafaelmartinelli** [comment]
So, I managed to make it work using EAV model (almost the same as my last message), and using the `CollectionField` from EasyAdminBundle. But it is quite ugly the way it is displayed:
![image](https://github.com/symfony/symfony/assets/10237929/290b2bb6-55e1-4d6a-8156-f13f4fc21a6d)
Now it's a question about EasyAdminBundle. Would it be possible to display those fields independently? I know how to create basic custom fields, but I don't know how to even start something like this... 😰
